File: platforms/android/app/src/main/java/com/ashera/constraintlayout/CircularFlowImpl.java
package com.ashera.constraintlayout;
//start - imports

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.ArrayList;

import android.content.Context;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.widget.*;
import androidx.core.view.*;
import android.view.*;
import android.graphics.drawable.*;


import android.os.Build;
import android.view.View;
import android.text.*;

import com.ashera.core.IFragment;
import com.ashera.converter.*;

import android.annotation.SuppressLint;

import com.ashera.layout.*;
import com.ashera.plugin.*;
import com.ashera.widget.bus.*;
import com.ashera.widget.*;
import com.ashera.widget.bus.Event.*;
import com.ashera.widget.IWidgetLifeCycleListener.EventId;
import com.ashera.widget.IWidgetLifeCycleListener.EventId.*;


import static com.ashera.widget.IWidget.*;
//end - imports
@SuppressLint("NewApi")
public class CircularFlowImpl extends BaseWidget {
	//start - body
	public final static String LOCAL_NAME = "androidx.constraintlayout.helper.widget.CircularFlow"; 
	public final static String GROUP_NAME = "androidx.constraintlayout.helper.widget.CircularFlow";

	protected androidx.constraintlayout.helper.widget.CircularFlow circularFlow;
	
	
	@Override
	public void loadAttributes(String attributeName) {
		ViewImpl.register(attributeName);


		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("circularflow_radiusInDP").withType("string"));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("circularflow_angles").withType("string"));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("circularflow_viewCenter").withType("id"));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("circularflow_defaultRadius").withType("dimension"));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("circularflow_defaultAngle").withType("float"));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("constraint_referenced_ids").withType("CircularFlowImpl.constraintReferencedIds"));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("addViewToCircularFlow").withType("object"));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("updateReference").withType("object"));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("updateRadius").withType("object"));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("updateAngle").withType("object"));
	}
	
	public CircularFlowImpl() {
		super(GROUP_NAME, LOCAL_NAME);
	}

		
	public class CircularFlowExt extends androidx.constraintlayout.helper.widget.CircularFlow implements ILifeCycleDecorator, com.ashera.widget.IMaxDimension{
		private MeasureEvent measureFinished = new MeasureEvent();
		private OnLayoutEvent onLayoutEvent = new OnLayoutEvent();
		private int mMaxWidth = -1;
		private int mMaxHeight = -1;
		@Override
		public void setMaxWidth(int width) {
			mMaxWidth = width;
		}
		@Override
		public void setMaxHeight(int height) {
			mMaxHeight = height;
		}
		@Override
		public int getMaxWidth() {
			return mMaxWidth;
		}
		@Override
		public int getMaxHeight() {
			return mMaxHeight;
		}

		public CircularFlowExt(Context context, android.util.AttributeSet attrs, int defStyleAttr) {
	        super(context, attrs, defStyleAttr);
	    }

		public CircularFlowExt(Context context) {
			super(context);
			
			
			
			
			
			
		}
		
		@Override
		public void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {

			if(mMaxWidth > 0) {
	        	widthMeasureSpec = MeasureSpec.makeMeasureSpec(mMaxWidth, MeasureSpec.AT_MOST);
	        }
	        if(mMaxHeight > 0) {
	            heightMeasureSpec = MeasureSpec.makeMeasureSpec(mMaxHeight, MeasureSpec.AT_MOST);

	        }

	        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
			IWidgetLifeCycleListener listener = (IWidgetLifeCycleListener) getListener();
			if (listener != null) {
			    measureFinished.setWidth(getMeasuredWidth());
			    measureFinished.setHeight(getMeasuredHeight());
				listener.eventOccurred(EventId.measureFinished, measureFinished);
			}
		}
		
		@Override
		protected void onLayout(boolean changed, int l, int t, int r, int b) {
			super.onLayout(changed, l, t, r, b);
			ViewImpl.nativeMakeFrame(asNativeWidget(), l, t, r, b);
			replayBufferedEvents();
			
			IWidgetLifeCycleListener listener = (IWidgetLifeCycleListener) getListener();
			if (listener != null) {
				onLayoutEvent.setB(b);
				onLayoutEvent.setL(l);
				onLayoutEvent.setR(r);
				onLayoutEvent.setT(t);
				onLayoutEvent.setChanged(changed);
				listener.eventOccurred(EventId.onLayout, onLayoutEvent);
			}
			
			if (isInvalidateOnFrameChange() && isInitialised()) {
				CircularFlowImpl.this.invalidate();
			}
		}	
		
		@Override
		public void onDraw(Canvas canvas) {
			Runnable runnable = () -> super.onDraw(canvas);
			executeMethodListeners("onDraw", runnable, canvas);
		}

		@Override
		public void draw(Canvas canvas) {
			Runnable runnable = () -> super.draw(canvas);
			executeMethodListeners("draw", runnable, canvas);
		}

		@SuppressLint("WrongCall")
		@Override
		public void execute(String method, Object... args) {
			switch (method) {
				case "onDraw":
					setOnMethodCalled(true);
					super.onDraw((Canvas) args[0]);
					break;

				case "draw":
					setOnMethodCalled(true);
					super.draw((Canvas) args[0]);
					break;

				default:
					break;
			}
		}

		public void updateMeasuredDimension(int width, int height) {
			setMeasuredDimension(width, height);
		}


		@Override
		public ILifeCycleDecorator newInstance(IWidget widget) {
			throw new UnsupportedOperationException();
		}

		@Override
		public void setAttribute(WidgetAttribute widgetAttribute,
				String strValue, Object objValue) {
			throw new UnsupportedOperationException();
		}		
		

		@Override
		public List<String> getMethods() {
			throw new UnsupportedOperationException();
		}
		
		@Override
		public void initialized() {
			throw new UnsupportedOperationException();
		}
		
        @Override
        public Object getAttribute(WidgetAttribute widgetAttribute) {
            throw new UnsupportedOperationException();
        }
        @Override
        public void drawableStateChanged() {
        	super.drawableStateChanged();
        	ViewImpl.drawableStateChanged(CircularFlowImpl.this);
        }
	}	
	public void updateMeasuredDimension(int width, int height) {
	    ((CircularFlowExt) circularFlow).updateMeasuredDimension(width, height);
	}

	@Override
	public IWidget newInstance() {
		return new CircularFlowImpl();
	}
	
	@SuppressLint("NewApi")
	@Override
	public void create(IFragment fragment, Map<String, Object> params) {
		super.create(fragment, params);
Context context = (Context) fragment.getRootActivity();
	Object systemStyle = params.get("systemStyle");
	Object systemAndroidAttrStyle = params.get("systemAndroidAttrStyle");
	
	if (systemStyle == null && systemAndroidAttrStyle == null) {
		circularFlow = new CircularFlowExt(context);
	} else {
		int defStyleAttr = 0;
		int defStyleRes = 0;
		
		if (systemStyle != null) {
			defStyleRes = context.getResources().getIdentifier((String) systemStyle, "style", context.getPackageName());	
		}
		
		if (systemAndroidAttrStyle != null) {
			defStyleAttr = context.getResources().getIdentifier((String) systemAndroidAttrStyle, "attr", "android");	
		}
		
		if (defStyleRes == 0) {
			circularFlow = new CircularFlowExt(context, null, defStyleAttr);	
		} else {
		}
		
	}

		nativeCreate(params);	
		ViewImpl.registerCommandConveter(this);
	}

	@Override
	@SuppressLint("NewApi")
	public void setAttribute(WidgetAttribute key, String strValue, Object objValue, ILifeCycleDecorator decorator) {
		Object nativeWidget = asNativeWidget();
		ViewImpl.setAttribute(this,  key, strValue, objValue, decorator);
		
		switch (key.getAttributeName()) {
			case "circularflow_radiusInDP": {
				


		 setRadiusInDP(objValue);



			}
			break;
			case "circularflow_angles": {
				


		 setAngles(objValue);



			}
			break;
			case "circularflow_viewCenter": {
				


		 setViewCenter(objValue);



			}
			break;
			case "circularflow_defaultRadius": {
				


	circularFlow.setDefaultRadius((int)objValue);



			}
			break;
			case "circularflow_defaultAngle": {
				


	circularFlow.setDefaultAngle((float)objValue);



			}
			break;
			case "constraint_referenced_ids": {
				


		 setConstraintReferenced_ids(objValue);



			}
			break;
			case "addViewToCircularFlow": {
				
		if (objValue instanceof Map) {
			Map<String, Object> data = ((Map<String, Object>) objValue);
		Object viewId = quickConvert(data.get("viewId"), "id");
		Object radius = quickConvert(data.get("radius"), "int");
		Object angle = quickConvert(data.get("angle"), "float");


		 addViewToCircularFlow ((int) viewId,(int)radius, (float)angle);


}
if (objValue instanceof java.util.List) {
	java.util.List<Object> list = (java.util.List<Object>) objValue;
	for (Object object : list) {
		Map<String, Object> data = PluginInvoker.getMap(object);
		Object viewId = quickConvert(data.get("viewId"), "id");
		Object radius = quickConvert(data.get("radius"), "int");
		Object angle = quickConvert(data.get("angle"), "float");


		 addViewToCircularFlow ((int) viewId,(int)radius, (float)angle);


	}
}
			}
			break;
			case "updateReference": {
				
		if (objValue instanceof Map) {
			Map<String, Object> data = ((Map<String, Object>) objValue);
		Object viewId = quickConvert(data.get("viewId"), "id");
		Object radius = quickConvert(data.get("radius"), "int");
		Object angle = quickConvert(data.get("angle"), "float");


		 updateReference ((int) viewId,(int)radius, (float)angle);


}
if (objValue instanceof java.util.List) {
	java.util.List<Object> list = (java.util.List<Object>) objValue;
	for (Object object : list) {
		Map<String, Object> data = PluginInvoker.getMap(object);
		Object viewId = quickConvert(data.get("viewId"), "id");
		Object radius = quickConvert(data.get("radius"), "int");
		Object angle = quickConvert(data.get("angle"), "float");


		 updateReference ((int) viewId,(int)radius, (float)angle);


	}
}
			}
			break;
			case "updateRadius": {
				
		if (objValue instanceof Map) {
			Map<String, Object> data = ((Map<String, Object>) objValue);
		Object viewId = quickConvert(data.get("viewId"), "id");
		Object radius = quickConvert(data.get("radius"), "int");


		 updateRadius ((int) viewId,(int)radius);


}
if (objValue instanceof java.util.List) {
	java.util.List<Object> list = (java.util.List<Object>) objValue;
	for (Object object : list) {
		Map<String, Object> data = PluginInvoker.getMap(object);
		Object viewId = quickConvert(data.get("viewId"), "id");
		Object radius = quickConvert(data.get("radius"), "int");


		 updateRadius ((int) viewId,(int)radius);


	}
}
			}
			break;
			case "updateAngle": {
				
		if (objValue instanceof Map) {
			Map<String, Object> data = ((Map<String, Object>) objValue);
		Object viewId = quickConvert(data.get("viewId"), "id");
		Object angle = quickConvert(data.get("angle"), "float");


		 updateAngle ((int) viewId, (float)angle);


}
if (objValue instanceof java.util.List) {
	java.util.List<Object> list = (java.util.List<Object>) objValue;
	for (Object object : list) {
		Map<String, Object> data = PluginInvoker.getMap(object);
		Object viewId = quickConvert(data.get("viewId"), "id");
		Object angle = quickConvert(data.get("angle"), "float");


		 updateAngle ((int) viewId, (float)angle);


	}
}
			}
			break;
		default:
			break;
		}
		postSetAttribute(key, strValue, objValue, decorator);
	}
	
	@Override
	@SuppressLint("NewApi")
	public Object getAttribute(WidgetAttribute key, ILifeCycleDecorator decorator) {
		Object nativeWidget = asNativeWidget();
		Object attributeValue = ViewImpl.getAttribute(this, nativeWidget, key, decorator);
		if (attributeValue != null) {
			return attributeValue;
		}
		switch (key.getAttributeName()) {
			case "constraint_referenced_ids": {
return getConstraintReferencedIds();				}
		}
		
		return null;
	}
	
	@Override
	public Object asWidget() {
		return circularFlow;
	}

	

    
    static {
        ConverterFactory.register("CircularFlowImpl.constraintReferencedIds", new ConstraintReferencedIdsConverter());
    }
    private void setConstraintReferenced_ids(Object objValue) {
        circularFlow.setReferencedIds((int[]) objValue);
    }
    
    private Object getConstraintReferencedIds() {
        return circularFlow.getReferencedIds();
    }
    

	
	    @Override
	    public Object asNativeWidget() {
	        return circularFlow;
	    }

	    private void nativeCreate(Map<String, Object> params) {
	    }
	@Override
	public void setId(String id){
		if (id != null && !id.equals("")){
			super.setId(id);
			circularFlow.setId(IdGenerator.getId(id));
		}
	}
	
 
    @Override
    public void requestLayout() {
    	if (isInitialised()) {
    		ViewImpl.requestLayout(this, asNativeWidget());
    		
    	}
    }
    
    @Override
    public void invalidate() {
    	if (isInitialised()) {
			ViewImpl.invalidate(this, asNativeWidget());

    	}
    }
    
	
private CircularFlowCommandBuilder builder;
private CircularFlowBean bean;
public Object getPlugin(String plugin) {
	return WidgetFactory.getAttributable(plugin).newInstance(this);
}
public CircularFlowBean getBean() {
	if (bean == null) {
		bean = new CircularFlowBean();
	}
	return bean;
}
public CircularFlowCommandBuilder getBuilder() {
	if (builder == null) {
		builder = new CircularFlowCommandBuilder();
	}
	return builder;
}


public  class CircularFlowCommandBuilder extends com.ashera.layout.ViewImpl.ViewCommandBuilder <CircularFlowCommandBuilder> {
    public CircularFlowCommandBuilder() {
	}
	
	public CircularFlowCommandBuilder execute(boolean setter) {
		if (setter) {
			executeCommand(command, null, IWidget.COMMAND_EXEC_SETTER_METHOD);
			getFragment().remeasure();
		}
		executeCommand(command, null, IWidget.COMMAND_EXEC_GETTER_METHOD);
return this;	}

public CircularFlowCommandBuilder setCircularflow_radiusInDP(String value) {
	Map<String, Object> attrs = initCommand("circularflow_radiusInDP");
	attrs.put("type", "attribute");
	attrs.put("setter", true);
	attrs.put("orderSet", ++orderSet);

	attrs.put("value", value);
return this;}
public CircularFlowCommandBuilder setCircularflow_angles(String value) {
	Map<String, Object> attrs = initCommand("circularflow_angles");
	attrs.put("type", "attribute");
	attrs.put("setter", true);
	attrs.put("orderSet", ++orderSet);

	attrs.put("value", value);
return this;}
public CircularFlowCommandBuilder setCircularflow_viewCenter(String value) {
	Map<String, Object> attrs = initCommand("circularflow_viewCenter");
	attrs.put("type", "attribute");
	attrs.put("setter", true);
	attrs.put("orderSet", ++orderSet);

	attrs.put("value", value);
return this;}
public CircularFlowCommandBuilder setCircularflow_defaultRadius(String value) {
	Map<String, Object> attrs = initCommand("circularflow_defaultRadius");
	attrs.put("type", "attribute");
	attrs.put("setter", true);
	attrs.put("orderSet", ++orderSet);

	attrs.put("value", value);
return this;}
public CircularFlowCommandBuilder setCircularflow_defaultAngle(float value) {
	Map<String, Object> attrs = initCommand("circularflow_defaultAngle");
	attrs.put("type", "attribute");
	attrs.put("setter", true);
	attrs.put("orderSet", ++orderSet);

	attrs.put("value", value);
return this;}
public CircularFlowCommandBuilder tryGetConstraint_referenced_ids() {
	Map<String, Object> attrs = initCommand("constraint_referenced_ids");
	attrs.put("type", "attribute");
	attrs.put("getter", true);
	attrs.put("orderGet", ++orderGet);
return this;}

public Object getConstraint_referenced_ids() {
	Map<String, Object> attrs = initCommand("constraint_referenced_ids");
	return attrs.get("commandReturnValue");
}
public CircularFlowCommandBuilder setConstraint_referenced_ids(String value) {
	Map<String, Object> attrs = initCommand("constraint_referenced_ids");
	attrs.put("type", "attribute");
	attrs.put("setter", true);
	attrs.put("orderSet", ++orderSet);

	attrs.put("value", value);
return this;}
public CircularFlowCommandBuilder addViewToCircularFlow(String viewId,
int radius,
float angle) {
	Map<String, Object> attrs = initCommand("addViewToCircularFlow");
	attrs.put("type", "attribute");
	attrs.put("setter", true);
	attrs.put("orderSet", ++orderSet);

	Map<String, Object> wrapper = new HashMap<>();
	wrapper.put("viewId", viewId);
	wrapper.put("radius", radius);
	wrapper.put("angle", angle);
	attrs.put("value", wrapper);
return this;}
public CircularFlowCommandBuilder updateReference(String viewId,
int radius,
float angle) {
	Map<String, Object> attrs = initCommand("updateReference");
	attrs.put("type", "attribute");
	attrs.put("setter", true);
	attrs.put("orderSet", ++orderSet);

	Map<String, Object> wrapper = new HashMap<>();
	wrapper.put("viewId", viewId);
	wrapper.put("radius", radius);
	wrapper.put("angle", angle);
	attrs.put("value", wrapper);
return this;}
public CircularFlowCommandBuilder updateRadius(String viewId,
int radius) {
	Map<String, Object> attrs = initCommand("updateRadius");
	attrs.put("type", "attribute");
	attrs.put("setter", true);
	attrs.put("orderSet", ++orderSet);

	Map<String, Object> wrapper = new HashMap<>();
	wrapper.put("viewId", viewId);
	wrapper.put("radius", radius);
	attrs.put("value", wrapper);
return this;}
public CircularFlowCommandBuilder updateAngle(String viewId,
float angle) {
	Map<String, Object> attrs = initCommand("updateAngle");
	attrs.put("type", "attribute");
	attrs.put("setter", true);
	attrs.put("orderSet", ++orderSet);

	Map<String, Object> wrapper = new HashMap<>();
	wrapper.put("viewId", viewId);
	wrapper.put("angle", angle);
	attrs.put("value", wrapper);
return this;}
}
public class CircularFlowBean extends com.ashera.layout.ViewImpl.ViewBean{
		public CircularFlowBean() {
			super(CircularFlowImpl.this);
		}
public void setCircularflow_radiusInDP(String value) {
	getBuilder().reset().setCircularflow_radiusInDP(value).execute(true);
}

public void setCircularflow_angles(String value) {
	getBuilder().reset().setCircularflow_angles(value).execute(true);
}

public void setCircularflow_viewCenter(String value) {
	getBuilder().reset().setCircularflow_viewCenter(value).execute(true);
}

public void setCircularflow_defaultRadius(String value) {
	getBuilder().reset().setCircularflow_defaultRadius(value).execute(true);
}

public void setCircularflow_defaultAngle(float value) {
	getBuilder().reset().setCircularflow_defaultAngle(value).execute(true);
}

public Object getConstraint_referenced_ids() {
	return getBuilder().reset().tryGetConstraint_referenced_ids().execute(false).getConstraint_referenced_ids(); 
}
public void setConstraint_referenced_ids(String value) {
	getBuilder().reset().setConstraint_referenced_ids(value).execute(true);
}

public void addViewToCircularFlow(String viewId,
int radius,
float angle) {
	getBuilder().reset().addViewToCircularFlow(viewId,
radius,
angle).execute(true);
}

public void updateReference(String viewId,
int radius,
float angle) {
	getBuilder().reset().updateReference(viewId,
radius,
angle).execute(true);
}

public void updateRadius(String viewId,
int radius) {
	getBuilder().reset().updateRadius(viewId,
radius).execute(true);
}

public void updateAngle(String viewId,
float angle) {
	getBuilder().reset().updateAngle(viewId,
angle).execute(true);
}

}


    
	//end - body


	private void setViewCenter(Object objValue) {
		setFieldUsingReflection(circularFlow, "mViewCenter", objValue);
	}
	
	private void setAngles(Object objValue) {
		initAngles();
		invokePrivateMethodUsingReflection(circularFlow, "setAngles", objValue);
	}
	
	private void setRadiusInDP(Object objValue) {
		initRadius();
		invokePrivateMethodUsingReflection(circularFlow, "setRadius", objValue);
	}
	
	private void postSetAttribute(WidgetAttribute key, String strValue, Object objValue,
			ILifeCycleDecorator decorator) {
		
		switch (key.getAttributeName()) {
		case "circularflow_defaultRadius": 
			setFieldUsingReflection(circularFlow, "mReferenceDefaultRadius", objValue);
			break;
		case "circularflow_defaultAngle": 
			setFieldUsingReflection(circularFlow, "mReferenceDefaultAngle", objValue);
			break;
		}
		if (isInitialised()) {
			invokePrivateMethodUsingReflection(circularFlow, "onAttachedToWindow");
		}
	}
	private void addViewToCircularFlow(int viewId, int radius, float angle) {
		circularFlow.addViewToCircularFlow((View) getParent().findWidgetById(IdGenerator.getName(viewId)).asWidget(), radius, angle);
	}

	private void initRadius() {
		try {
			circularFlow.getRadius();
		} catch (Exception e) {
			setFieldUsingReflection(circularFlow, "mRadius", new int[1]);
		}
	}

	private void initAngles() {
		try {
			circularFlow.getAngles();
		} catch (Exception e) {
			setFieldUsingReflection(circularFlow, "mAngles", new float[1]);
		}
	}
	
	private void updateAngle(int viewId, float angle) {
		circularFlow.updateAngle((View) getParent().findWidgetById(IdGenerator.getName(viewId)).asWidget(), angle);
		
	}

	private void updateRadius(int viewId, int radius) {
		circularFlow.updateRadius((View) getParent().findWidgetById(IdGenerator.getName(viewId)).asWidget(), radius);
		
	}

	private void updateReference(int viewId, int radius, float angle) {
		circularFlow.updateReference((View) getParent().findWidgetById(IdGenerator.getName(viewId)).asWidget(), radius, angle);
		
	}


}

File: platforms/android/app/src/main/tsc/src/android/widget/CircularFlowImpl.ts
// start - imports

	
import CommandAttr from '../../widget/CommandAttr';
import IWidget from '../../widget/IWidget';
import ILayoutParam from '../../widget/ILayoutParam';
import {plainToClass, Type, Exclude, Expose, Transform} from "class-transformer";
import 'babel-polyfill';
import {Gravity} from '../../widget/TypeConstants';
import {ITranform, TransformerFactory} from '../../widget/TransformerFactory';
import {Event} from '../../app/Event';
import {MotionEvent} from '../../app/MotionEvent';
import {DragEvent} from '../../app/DragEvent';
import {KeyEvent} from '../../app/KeyEvent';
import { ScopedObject } from '../../app/ScopedObject';







export class CircularFlowImpl_addViewToCircularFlow {
@Expose({ name: "viewId" })
viewId!:string;
@Expose({ name: "radius" })
radius!:number;
@Expose({ name: "angle" })
angle!:number;
}
export class CircularFlowImpl_updateReference {
@Expose({ name: "viewId" })
viewId!:string;
@Expose({ name: "radius" })
radius!:number;
@Expose({ name: "angle" })
angle!:number;
}
export class CircularFlowImpl_updateRadius {
@Expose({ name: "viewId" })
viewId!:string;
@Expose({ name: "radius" })
radius!:number;
}
export class CircularFlowImpl_updateAngle {
@Expose({ name: "viewId" })
viewId!:string;
@Expose({ name: "angle" })
angle!:number;
}


// end - imports
import {ViewImpl} from './ViewImpl';
export abstract class CircularFlowImpl<T> extends ViewImpl<T>{
	//start - body
	static initialize() {
    }	
	@Type(() => CommandAttr)
	@Expose({ name: "circularflow_radiusInDP" })
	circularflow_radiusInDP!:CommandAttr<string>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "circularflow_angles" })
	circularflow_angles!:CommandAttr<string>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "circularflow_viewCenter" })
	circularflow_viewCenter!:CommandAttr<string>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "circularflow_defaultRadius" })
	circularflow_defaultRadius!:CommandAttr<string>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "circularflow_defaultAngle" })
	circularflow_defaultAngle!:CommandAttr<number>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "constraint_referenced_ids" })
	constraint_referenced_ids!:CommandAttr<string>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "addViewToCircularFlow" })
	addViewToCircularFlow_!:CommandAttr<CircularFlowImpl_addViewToCircularFlow>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "updateReference" })
	updateReference_!:CommandAttr<CircularFlowImpl_updateReference>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "updateRadius" })
	updateRadius_!:CommandAttr<CircularFlowImpl_updateRadius>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "updateAngle" })
	updateAngle_!:CommandAttr<CircularFlowImpl_updateAngle>| undefined;

	@Exclude()
	protected thisPointer: T;	
	protected abstract getThisPointer(): T;
	reset() : T {	
		super.reset();
		this.circularflow_radiusInDP = undefined;
		this.circularflow_angles = undefined;
		this.circularflow_viewCenter = undefined;
		this.circularflow_defaultRadius = undefined;
		this.circularflow_defaultAngle = undefined;
		this.constraint_referenced_ids = undefined;
		this.addViewToCircularFlow_ = undefined;
		this.updateReference_ = undefined;
		this.updateRadius_ = undefined;
		this.updateAngle_ = undefined;
		return this.thisPointer;
	}
	constructor(id: string, path: string[], event:  string) {
		super(id, path, event);
		this.thisPointer = this.getThisPointer();
	}
	

	public setCircularflow_radiusInDP(value : string) : T {
		this.resetIfRequired();
		if (this.circularflow_radiusInDP == null || this.circularflow_radiusInDP == undefined) {
			this.circularflow_radiusInDP = new CommandAttr<string>();
		}
		
		this.circularflow_radiusInDP.setSetter(true);
		this.circularflow_radiusInDP.setValue(value);
		this.orderSet++;
		this.circularflow_radiusInDP.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public setCircularflow_angles(value : string) : T {
		this.resetIfRequired();
		if (this.circularflow_angles == null || this.circularflow_angles == undefined) {
			this.circularflow_angles = new CommandAttr<string>();
		}
		
		this.circularflow_angles.setSetter(true);
		this.circularflow_angles.setValue(value);
		this.orderSet++;
		this.circularflow_angles.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public setCircularflow_viewCenter(value : string) : T {
		this.resetIfRequired();
		if (this.circularflow_viewCenter == null || this.circularflow_viewCenter == undefined) {
			this.circularflow_viewCenter = new CommandAttr<string>();
		}
		
		this.circularflow_viewCenter.setSetter(true);
		this.circularflow_viewCenter.setValue(value);
		this.orderSet++;
		this.circularflow_viewCenter.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public setCircularflow_defaultRadius(value : string) : T {
		this.resetIfRequired();
		if (this.circularflow_defaultRadius == null || this.circularflow_defaultRadius == undefined) {
			this.circularflow_defaultRadius = new CommandAttr<string>();
		}
		
		this.circularflow_defaultRadius.setSetter(true);
		this.circularflow_defaultRadius.setValue(value);
		this.orderSet++;
		this.circularflow_defaultRadius.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public setCircularflow_defaultAngle(value : number) : T {
		this.resetIfRequired();
		if (this.circularflow_defaultAngle == null || this.circularflow_defaultAngle == undefined) {
			this.circularflow_defaultAngle = new CommandAttr<number>();
		}
		
		this.circularflow_defaultAngle.setSetter(true);
		this.circularflow_defaultAngle.setValue(value);
		this.orderSet++;
		this.circularflow_defaultAngle.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public tryGetConstraint_referenced_ids() : T {
		this.resetIfRequired();
		if (this.constraint_referenced_ids == null || this.constraint_referenced_ids == undefined) {
			this.constraint_referenced_ids = new CommandAttr<string>()
		}
		
		this.constraint_referenced_ids.setGetter(true);
		this.orderGet++;
		this.constraint_referenced_ids.setOrderGet(this.orderGet);
		return this.thisPointer;
	}
	
	public getConstraint_referenced_ids() : string {
		if (this.constraint_referenced_ids == null || this.constraint_referenced_ids == undefined) {
			this.constraint_referenced_ids = new CommandAttr<string>();
		}
		return this.constraint_referenced_ids.getCommandReturnValue();
	}
	public setConstraint_referenced_ids(value : string) : T {
		this.resetIfRequired();
		if (this.constraint_referenced_ids == null || this.constraint_referenced_ids == undefined) {
			this.constraint_referenced_ids = new CommandAttr<string>();
		}
		
		this.constraint_referenced_ids.setSetter(true);
		this.constraint_referenced_ids.setValue(value);
		this.orderSet++;
		this.constraint_referenced_ids.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public addViewToCircularFlow(viewId : string,
radius : number,
angle : number) : T {
		this.resetIfRequired();
		if (this.addViewToCircularFlow_ == null || this.addViewToCircularFlow_ == undefined) {
			this.addViewToCircularFlow_ = new CommandAttr<CircularFlowImpl_addViewToCircularFlow>();
		}
		
		let wrapper:CircularFlowImpl_addViewToCircularFlow = new CircularFlowImpl_addViewToCircularFlow();
		wrapper.viewId = viewId;
		wrapper.radius = radius;
		wrapper.angle = angle;
		this.addViewToCircularFlow_.setSetter(true);
		this.addViewToCircularFlow_.setValue(wrapper);	
		this.orderSet++;
		this.addViewToCircularFlow_.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public updateReference(viewId : string,
radius : number,
angle : number) : T {
		this.resetIfRequired();
		if (this.updateReference_ == null || this.updateReference_ == undefined) {
			this.updateReference_ = new CommandAttr<CircularFlowImpl_updateReference>();
		}
		
		let wrapper:CircularFlowImpl_updateReference = new CircularFlowImpl_updateReference();
		wrapper.viewId = viewId;
		wrapper.radius = radius;
		wrapper.angle = angle;
		this.updateReference_.setSetter(true);
		this.updateReference_.setValue(wrapper);	
		this.orderSet++;
		this.updateReference_.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public updateRadius(viewId : string,
radius : number) : T {
		this.resetIfRequired();
		if (this.updateRadius_ == null || this.updateRadius_ == undefined) {
			this.updateRadius_ = new CommandAttr<CircularFlowImpl_updateRadius>();
		}
		
		let wrapper:CircularFlowImpl_updateRadius = new CircularFlowImpl_updateRadius();
		wrapper.viewId = viewId;
		wrapper.radius = radius;
		this.updateRadius_.setSetter(true);
		this.updateRadius_.setValue(wrapper);	
		this.orderSet++;
		this.updateRadius_.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public updateAngle(viewId : string,
angle : number) : T {
		this.resetIfRequired();
		if (this.updateAngle_ == null || this.updateAngle_ == undefined) {
			this.updateAngle_ = new CommandAttr<CircularFlowImpl_updateAngle>();
		}
		
		let wrapper:CircularFlowImpl_updateAngle = new CircularFlowImpl_updateAngle();
		wrapper.viewId = viewId;
		wrapper.angle = angle;
		this.updateAngle_.setSetter(true);
		this.updateAngle_.setValue(wrapper);	
		this.orderSet++;
		this.updateAngle_.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		
	//end - body

}
	
//start - staticinit

export class CircularFlow extends CircularFlowImpl<CircularFlow> implements IWidget{
    getThisPointer(): CircularFlow {
        return this;
    }
    
   	public getClass() {
		return CircularFlow;
	}
	
   	constructor(id: string, path: string[], event: string) {
		super(id, path, event);	
	}
}

CircularFlowImpl.initialize();

//end - staticinit
